File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
from collections import deque

n, m = map(int, input().split())

miro = [[] for _ in range(n)]

for i in range(n):
    s = input()
    for j in s:
        miro[i].append(int(j))

dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]

def bfs():
    d = deque()
    d.append((0, 0))

    while d:
        x, y = d.popleft()

        for i in range(4):
            newX = x + dx[i]
            newY = y + dy[i]

            if newX < 0 or newX >= n:
                continue
            if newY < 0 or newY >= m:
                continue
            if miro[newX][newY] != 1:
                continue
            
            miro[newX][newY] = miro[x][y] + 1
            d.append((newX, newY))
    
    return miro[n-1][m-1]

print(bfs())

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int[][] miro;
    static int[] dx = {0, 0, -1, 1};
    static int[] dy = {-1, 1, 0, 0};
    static int n, m;
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer stringTokenizer = new StringTokenizer(bufferedReader.readLine());

        n = Integer.parseInt(stringTokenizer.nextToken());
        m = Integer.parseInt(stringTokenizer.nextToken());
        miro = new int[n][m];

        for (int i = 0; i < n; i++) {
            String input = bufferedReader.readLine();
            for (int j = 0; j < m; j++) {
                miro[i][j] = input.charAt(j) - '0';
            }
        }

        bfs();

        System.out.println(miro[n-1][m-1]);
    }

    static void bfs() {
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{0, 0});

        while (!queue.isEmpty()) {
            int[] array = queue.poll();
            int x = array[0];
            int y = array[1];

            for (int i = 0; i < 4; i++) {
                int newX = x + dx[i];
                int newY = y + dy[i];

                if (0 <= newX && newX < n && 0 <= newY && newY < m) {
                    if (miro[newX][newY] != 1) continue;
                    miro[newX][newY] = miro[x][y] + 1;
                    queue.add(new int[]{newX, newY});
                }
            }
        }
    }
}
